crates/glaredb/src/lib.rs

```diff
@@ -0,0 +1,396 @@
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::path::PathBuf;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use datafusion::arrow::array::{StringArray, UInt64Array};
+use datafusion::arrow::datatypes::{DataType, Field, Schema};
+// public re-export so downstream users of this package don't have to
+// directly depend on DF (and our version no-less) to use our interfaces.
+pub use datafusion::arrow::record_batch::RecordBatch;
+pub use datafusion::error::DataFusionError;
+use datafusion::logical_expr::LogicalPlan;
+use datafusion::physical_plan::stream::RecordBatchStreamAdapter;
+pub use datafusion::physical_plan::SendableRecordBatchStream;
+use derive_builder::Builder;
+use futures::lock::Mutex;
+use futures::stream::{Stream, StreamExt};
+use sqlexec::engine::{Engine, EngineBackend, TrackedSession};
+use sqlexec::environment::EnvironmentReader;
+use sqlexec::errors::ExecError;
+use sqlexec::remote::client::RemoteClientType;
+use sqlexec::session::ExecutionResult;
+use sqlexec::OperationInfo;
+use url::Url;
+
+
+#[derive(Default, Builder)]
+pub struct ConnectOptions {
+    #[builder(setter(into))]
+    pub connection_target: Option<String>,
+    #[builder(setter(into))]
+    pub location: Option<String>,
+    #[builder(setter(into))]
+    pub spill_path: Option<String>,
+    #[builder(setter(into), default = "HashMap::new()")]
+    pub storage_options: HashMap<String, String>,
+
+    #[builder(setter(strip_option))]
+    pub disable_tls: Option<bool>,
+    #[builder(default = "Some(\"https://console.glaredb.com\".to_string())")]
+    #[builder(setter(into, strip_option))]
+    pub cloud_addr: Option<String>,
+    #[builder(default = "Some(RemoteClientType::Cli)")]
+    #[builder(setter(strip_option))]
+    pub client_type: Option<RemoteClientType>,
+    #[builder(setter(strip_option))]
+    pub environment_reader: Option<Arc<Box<dyn EnvironmentReader>>>,
```

Comment:
otherwise it's not cloneable, and in order to pass the same environment reader from the language environment, to the builder/options struct to the session it needs to be cloneable. Previously we just constructed the python env reader during connection.

I tried removing the arc, and I suppose we could pass it in without an arc to the connection object but that feels like it would happen "late", and the Arc seems clean.